providers/terraform-provider-csbpg/csbpg/connection_factory.go

```diff
@@ -21,20 +21,24 @@ type connectionFactory struct {
 }
 
 func (c connectionFactory) Connect() (*sql.DB, error) {
-	db, err := sql.Open("postgres", c.uri())
+	return c.ConnectWithURI(c.uri())
+}
+
+func (c connectionFactory) ConnectWithURI(uri string) (*sql.DB, error) {
+	db, err := sql.Open("postgres", uri)
 	if err != nil {
 		return nil, fmt.Errorf("failed to connect to PostgreSQL %q: %w", c.uriRedacted(), err)
 	}
 
 	return db, nil
 }
 
-func (c connectionFactory) uri() string {
+func (c connectionFactory) URIWithCreds(username, password string) string {
```

Comment:
This API seems a little messy. I can see how it evolved, but I'd prefer `ConnectAsAdmin()` and `ConnectAsUser(username, password string)` or something that helps with understanding.